clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp

```diff
@@ -709,6 +709,145 @@ class CIRGetGlobalOpLowering
   }
 };
 
+class CIRCastOpLowering : public mlir::OpConversionPattern<mlir::cir::CastOp> {
+public:
+  using OpConversionPattern<mlir::cir::CastOp>::OpConversionPattern;
+
+  inline mlir::Type convertTy(mlir::Type ty) const {
+    return getTypeConverter()->convertType(ty);
+  }
+
+  /// If the given type is a vector type, return the vector's element type.
+  /// Otherwise return the given type unchanged.
+  inline mlir::Type elementTypeIfVector(mlir::Type type) const {
+    if (auto VecType = type.dyn_cast<mlir::cir::VectorType>()) {
+      return VecType.getEltType();
+    }
+    return type;
+  }
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::CastOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    auto src = adaptor.getSrc();
+    auto dstType = op.getResult().getType();
+    using CIR = mlir::cir::CastKind;
+    switch (op.getKind()) {
+    case CIR::int_to_bool: {
+      auto zero = rewriter.create<mlir::cir::ConstantOp>(
+          src.getLoc(), op.getSrc().getType(),
+          mlir::cir::IntAttr::get(op.getSrc().getType(), 0));
+      rewriter.replaceOpWithNewOp<mlir::cir::CmpOp>(
+          op, mlir::cir::BoolType::get(getContext()), mlir::cir::CmpOpKind::ne,
+          op.getSrc(), zero);
+      return mlir::success();
+    }
+    case CIR::integral: {
+      auto srcType = op.getSrc().getType();
+      auto newDstType = convertTy(dstType);
+      mlir::cir::IntType srcIntType =
+          elementTypeIfVector(srcType).cast<mlir::cir::IntType>();
+      mlir::cir::IntType dstIntType =
+          elementTypeIfVector(dstType).cast<mlir::cir::IntType>();
+
+      if (dstIntType.getWidth() < srcIntType.getWidth()) {
```

Comment:
Please refactor this logic into something similar to `getLLVMIntCast` in `LowerToLLVM.cpp`, so it can be reused next time MLIR lowering requires trunc/extensions.

Reply:
It's a good idea.